crates/ruff_python_formatter/src/expression/expr_bin_op.rs

```diff
@@ -179,42 +233,94 @@ impl NeedsParentheses for ExprBinOp {
+                {
                     Parentheses::Optional
+                } else {
+                    Parentheses::Custom
                 }
             }
             parentheses => parentheses,
         }
     }
 }
 
-pub(super) fn should_binary_break_right_side_first(expr: &ExprBinOp) -> bool {
-    use ruff_python_ast::prelude::*;
+#[derive(Copy, Clone, Debug, Eq, PartialEq)]
+enum BinaryLayout {
+    /// Put each operand on their own line if either side expands
+    Default,
 
-    if expr.left.is_bin_op_expr() {
-        false
-    } else {
-        match expr.right.as_ref() {
-            Expr::Tuple(ExprTuple {
-                elts: expressions, ..
-            })
-            | Expr::List(ExprList {
-                elts: expressions, ..
-            })
-            | Expr::Set(ExprSet {
-                elts: expressions, ..
-            })
-            | Expr::Dict(ExprDict {
-                values: expressions,
-                ..
-            }) => !expressions.is_empty(),
-            Expr::Call(ExprCall { args, keywords, .. }) => !args.is_empty() && !keywords.is_empty(),
-            Expr::ListComp(_) | Expr::SetComp(_) | Expr::DictComp(_) | Expr::GeneratorExp(_) => {
-                true
-            }
-            _ => false,
+    /// Try to expand the left to make it fit. Add parentheses if the left or right don't fit.
+    ///
+    ///```python
+    /// [
+    ///     a,
+    ///     b
+    /// ] & c
+    ///```
+    ExpandLeft,
+
+    /// Try to expand the right to make it fix. Add parentheses if the left or right don't fit.
+    ///
+    /// ```python
+    /// a & [
+    ///     b,
+    ///     c
+    /// ]
+    /// ```
+    ExpandRight,
+
+    /// Both the left and right side can be expanded. Try in the following order:
+    /// * expand the right side
+    /// * expand the left side
+    /// * expand both sides
+    ///
+    /// to make the expression fit
+    ///
+    /// ```python
+    /// [
+    ///     a,
+    ///     b
+    /// ] & [
+    ///     c,
+    ///     d
+    /// ]
+    /// ```
+    ExpandRightThenLeft,
+}
+
+impl BinaryLayout {
+    fn from(expr: &ExprBinOp) -> Self {
+        match (can_break(&expr.left), can_break(&expr.right)) {
+            (false, false) => Self::Default,
+            (true, false) => Self::ExpandLeft,
+            (false, true) => Self::ExpandRight,
+            (true, true) => Self::ExpandRightThenLeft,
         }
     }
 }
+
+fn can_break(expr: &Expr) -> bool {
+    use ruff_python_ast::prelude::*;
+
+    match expr {
+        Expr::Tuple(ExprTuple {
+            elts: expressions, ..
+        })
+        | Expr::List(ExprList {
+            elts: expressions, ..
+        })
+        | Expr::Set(ExprSet {
+            elts: expressions, ..
+        })
+        | Expr::Dict(ExprDict {
+            values: expressions,
+            ..
+        }) => !expressions.is_empty(),
+        Expr::Call(ExprCall { args, keywords, .. }) => !args.is_empty() && !keywords.is_empty(),
```

Comment:
Shouldn't this be `||`, e.g. we just need any kind of arguments to allow breaking?

Reply:
Good catch, thanks